### PR TITLE
Implement a lot of clippy suggestions

### PR DESCRIPTION
**Pull request by MajorBreakfast**

### CODE REVIEW

File: futures-util/src/future/join.rs [hunk 1]
```diff
@@ -57,7 +59,7 @@ macro_rules! generate {
                 )*
 
                 if all_done {
-                    Poll::Ready(($(self.$Fut().take_output().unwrap_or_else(|| unreachable!())),*))
```
Comment (MajorBreakfast) [inline futures-util/src/future/join.rs]: This looks strange, but at the same time it feels like I'm missing something

File: futures-util/src/stream/unfold.rs [hunk 2]
```diff
@@ -90,21 +90,19 @@ impl<T, F, Fut, It> Stream for Unfold<T, F, Fut>
...
-                *self.state() = Some(next_state);
-                return Poll::Ready(Some(item))
-            } else {
-                return Poll::Ready(None)
-            }
+        if let Some((item, next_state)) = step {
+            *self.state() = Some(next_state);
+            return Poll::Ready(Some(item))
+        } else {
+            return Poll::Ready(None)
```
Comment (MajorBreakfast) [inline futures-util/src/stream/unfold.rs]: Here clippy said the loop was useless.

File: futures-util/src/future/join.rs (hunk 1 above)
Comment (cramertj) [reply futures-util/src/future/join.rs]: These aren't quite the same-- `unwrap` only works for `Debug` types, while `unwrap_or_else(|| unreachable!())` works for all types.

Comment (cramertj) [reply futures-util/src/future/join.rs]: Edit: ignore last-- that used to be true when this was a `Result`, but now that it is an `Option`, unwrapping is fine.

File: futures-util/src/stream/concat.rs [hunk 3]
```diff
@@ -46,17 +46,21 @@ impl<S: Stream + Unpin> Unpin for Concat<S> {}
...
+    fn poll(
+        mut self: PinMut<Self>, cx: &mut task::Context
+    ) -> Poll<Self::Output> {
         loop {
             match self.stream().poll_next(cx) {
                 Poll::Pending => return Poll::Pending,
                 Poll::Ready(None) => {
-                    return Poll::Ready(self.accum().take().unwrap_or_else(|| Default::default()))
+                    return Poll::Ready(
+                        self.accum().take().unwrap_or_else(Default::default)
```
Comment (cramertj) [inline futures-util/src/stream/concat.rs]: Even better: swap out for `unwrap_or_default`: https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_default

File: futures-util/src/stream/unfold.rs (hunk 2 above)
Comment (cramertj) [reply futures-util/src/stream/unfold.rs]: Nice! Looks like an artifact of how this used to be written: the `self.state`-setting case would wrap around: https://github.com/rust-lang-nursery/futures-rs/blob/master/futures-util/src/stream/unfold.rs#L89

**Comment by cramertj** [comment]
Cool! Thanks for these fixes.

File: futures-util/src/stream/concat.rs (hunk 3 above)
Comment (MajorBreakfast) [reply futures-util/src/stream/concat.rs]: cool. Now it will fit into one line

Comment (MajorBreakfast) [reply futures-util/src/stream/concat.rs]: I'll simply include this in the next PR

**Pull request merged**